### ISSUE TITLE
[Device Support Request for HG0646]

### ISSUE THREAD

**Issue by erkr** [zigpy/zha-device-handlers#1933]
Please add a quirk to support the melinera HG06467 Christmas light strings sold by the LIDL.

The strings are supported by Z2M: https://www.zigbee2mqtt.io/devices/HG06467.html

And there exists an issue for this that became stale. That issue contains all information wrt these Christmas led strings: 
https://github.com/zigpy/zha-device-handlers/issues/604

Best 
Eric

**Comment by MattWestb** [comment]
Hola @javicalle  !!

If you have time this device is very wanted for the moment and is one real "tuya hybrid" that is using normal Zigbee commands and some very tuya cooked ones.

In ground is one normal light that have pixel LEDs and the normal lights things is Zigbe commands and the pixel is having some special commands implanted.

I have not baying it then is was little over my budged but many users have doing it.

If you have time and can taking one look may user wold getting one nice Christmases if you can cocking one quirk for it.

Mvh Mattias

**Comment by rkruisselbrink** [comment]
Is there anything we can do to help with this effort? I have this product as well and it would be awesome to control the effects via ZHA 🎄

**Comment by erkr** [comment]
I can help to test with local quirks

**Comment by erkr** [comment]
A lot of time left before Christmas 😉

**Comment by MattWestb** [comment]
@TheJulianJES We was trying getting one of this to our US friends for2 years ago but we was not getting any that was willing baying and shipping it from Europe.
If you is getting one or some else in CE is having one and can getting it to you can you taking one look getting one working quirk for it ?
I think its one normal RGB light plus extra scenes for searing the pixel LEDs with tuya MCU commands.

**Comment by MattWestb** [comment]
For users that having one of this "devices" and need it working next Christmas is the devs needing getting hand of one for testing and doing the coding.

In CE we have great quirk devs in Germany and in Spain and probably more if we asking little more around.

Pleas helping getting the beloved device working by sharing hardware for some time and getting it working with our systems.

Thanks in advance !!!

**Comment by xanibale** [comment]
I am from Germany and have this Device, i would be willing to send it to any quirk dev, preferable in Germany, for some time.

**Comment by MattWestb** [comment]
Great @xanibale I pinging @TheJulianJES hi is located in Ger and have the knowledge and if need hi is getting help from our US or Spanish devs.

I hoping for the next Christmas shall being white with much colors in the (dev)tree !!!

**Comment by TheJulianJES** [comment]
At the moment, I don't have enough time to work on a (Tuya) device with a given time limit.

It'll also be hard to implement the effects properly and ZHA (HA Core) would probably also need to be modified for this.

**Comment by xanibale** [comment]
If i would get it back until next Christmas, then there should be no problem, as it is now not usable for me anyway.
It would be very nice to use the colour/ white capabilities of the device. If that would make things easier.

**Comment by erkr** [comment]
Yes, first fixing normal colour and brightness control would be a great step. Now the colours change when the brightness is adjusted.

**Comment by erkr** [comment]
Christmas seem still far away, but we are 6 months further 🙏

**Comment by rkruisselbrink** [comment]
I was also thinking about this. Is there some guide or docs on how we can test/debug this? I'm willing to give it a shot and spent some time on this, but i have no idea where to start 😅

**Comment by d3nnis0** [comment]
I also own the those Christmas lights and I'am willing to invest some time to get them integrated until Christmas (Or even better some weeks before ;-) ) 

Is there already a kind of draft where I can try to contribute?

**Comment by rkruisselbrink** [comment]
> Is there already a kind of draft where I can try to contribute?

Not that i know of. Any idea how to approach this?

**Comment by erkr** [comment]
I noticed that the Christmas lights where added to the  Deconz integration: https://github.com/Kane610/deconz/issues/198

I hope some expert knows how to translate the information that seems to be available there into a handler for ZHA

**Comment by Hedda** [comment]
> I noticed that the Christmas lights where added to the Deconz integration: [Kane610/deconz#198](https://github.com/Kane610/deconz/issues/198)
> 
> I hope some expert knows how to translate the information that seems to be available there into a handler for ZHA

FYI, for development reference, the developer of the deCONZ integration for Home Assistant also added "effects" and "colors" support for HG06467 (Zigbee ID "TS0601") Lidl Silvercrest Christmas light to Home Assistant's deCONZ integration now too:

https://github.com/home-assistant/core/pull/102731 -> https://github.com/home-assistant/core/pull/102731/commits/d58dcd179ec322602afb74ee909bde93fbbdf022

https://github.com/home-assistant/core/pull/105155 -> https://github.com/home-assistant/core/pull/105155/commits/fecdb6c20fd78d45007a483e76920bfbb3a27ed2

https://community.home-assistant.io/t/lidl-melinera-christmas-string-hg06467-lidl-livarno-lux/367

https://community.home-assistant.io/t/update-deconz-integration-to-support-flash-with-color/246904/

PS: For more reference, same "effect" and "color" support for Lidl Malinera "Smart LED String Lights" HG06467 (Zigbee ID "TS0601") was added to zigbee-herdsman-converters for Zigbee2MQTT back in 2020 and these are their pull request and discussions about it:

https://github.com/Koenkk/zigbee-herdsman-converters/pull/1799

https://github.com/Koenkk/zigbee-herdsman-converters/issues/1792

https://zigbee2mqtt.discourse.group/t/lidl-christmas-lights/1788/

https://github.com/Koenkk/zigbee2mqtt/issues/5221

**Comment by enyone** [comment]
I tried to hack something together last year this same time (one week before Christmas). My neighbors are still not happy so also I am keen to debug any contributions to this draft.

https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6

See that `0xFC: foundation.ZCLCommandDef(...` I created to issue effect change command
and a handler for that `if command_id == 0x00FC:`
with a static data `tuya_payload.data = [0x00, 0x0A, 0x30, 0x31, 0x30, 0x30, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40]` to play around.

**Comment by rkruisselbrink** [comment]
I think we want to keep this open. I didn't have (make) time yet to checkout the implementation by @enyone

**Comment by Melantrix** [comment]
> I tried to hack something together last year this same time (one week before Christmas). My neighbors are still not happy so also I am keen to debug any contributions to this draft.
> 
> https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6
> 
> See that `0xFC: foundation.ZCLCommandDef(...` I created to issue effect change command and a handler for that `if command_id == 0x00FC:` with a static data `tuya_payload.data = [0x00, 0x0A, 0x30, 0x31, 0x30, 0x30, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40]` to play around.

I don't know if you've looked at this some more, but thanks to this i'm able to choose colors, however the colors i choose don't seem to correspond to the colors i choose (e.g. i choose red in HA, but the light turn yellow). it seems that color temperature is not set correctly. I'm able/willing to help troubleshoot/debug but i have no i idea what i'm looking at in the quirk.. :/ Can you maybe point me in the right direction?

**Comment by enyone** [comment]
I still have these lights on my yard's bushes.

For now I've beard with that broken XY-mode, where actual string's LED color is not always quite the one one picks at RGB map palette.

This https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6 actually brings only (sightly broken) XY-mode available (where you can pick one led color for all string's LED using RGB map palette). Without that quirk only HS-mode is available (using Hue slider and/or buttons) but it does not work at all. For some reason this device reports HS-mode as only supported mode, whereas that quirk forces the modes XY-mode and CT-mode as supported ones. See `LidlRGBColorCluster` in quirk code.

Why XY-mode is broken, I don't know. I would guess this device uses some non-8-bit (but less) color depth.

Another thing that quirk tries to create support for is effects (flashing multi-color effects Lidl's Android App specifies). I have not found any documentation or similar implementation elsewhere to use as a reference, and fall back to reverse-engineering bytes using Lidl's Android App as a source of those bytes. One example set of bytes is present in that quirk code line 98:
`tuya_payload.data = [0x00, 0x0A, 0x30, 0x31, 0x30, 0x30, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40]`

That array of bytes is send as a payload data to device using `TuyaManufCluster.Command` with:
`command_id=TUYA_SET_DATA (0x00)`
`function=TUYA_EFFECT_COMMAND (0x0603)`

See `LidlEffectCluster` in quirk code and more from https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/__init__.py#L324

That array of bytes is just an example and more different samples (by reverse-engineering Lidl's Android App sent payloads) needs to be observed to create something working.

**Comment by erkr** [comment]
z2m seems to support the light effects. So someone knowledgeable can reuse that for zha I guess

**Comment by smoltron** [comment]
This seems to be broken in ZHA now. Last year it still kind of worked, I could control color. Now only on-off works as expected. The color setting/brightness changes colors randomly.

**Comment by erkr** [comment]
Plenty of time left till next Christmas 🤓

**Comment by TjakBak** [comment]
Ho Ho Ho, Christmas is around the corner again, and Santa asked me to revive this 3-year-old thread to see if anyone following has managed to fix the quirk and get the effects selection working:)

Or is it officially time to throw these out and buy new ones? 

Anyway, Happy holidays everyone!!!

**Comment by smoltron** [comment]
If one uses a device one week a year, it will take so many years until the device is done. Therefore, I would like to reuse this set as long as possible. It is also ecological.

**Comment by smoltron** [comment]
ZHA cannot control the light color at all. Is there a solution?

**Comment by hucki** [comment]
> ZHA cannot control the light color at all. Is there a solution?

Hey. My solution was to switch from ZHA to zigbee2mqtt yesterday.

**Comment by smoltron** [comment]
Yes, I thought of that. However, this custom quirk, https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6 seems to mostly work, at least sufficiently. Put it to config/custom_zha_quirks/TS0601.py

**Comment by erkr** [comment]
Yes @enyone example quirk works really nice to control it as a light 👌
As mentioned above, control for the effects is missing. This log seems related:
``` 

Logger: zigpy.zcl
Source: custom_quirks/LIDL melinera.py:56
First occurred: 18:29:50 (1 occurrence)
Last logged: 18:29:50

Command 'effect' has an incorrect direction, please remove the `direction` kwarg

``` 

I don't see a direction in the code though

**Comment by Hedda** [comment]
FYI, for reference see https://github.com/zigpy/zigpy/issues/1640

- https://github.com/zigpy/zigpy/issues/1640

**Comment by enyone** [comment]
I still have these but my neighbor does not like blinking effects. Thus I must have been downgraded my expectations of ZHA support of this device and done nothing new.

**Comment by erkr** [comment]
@enyone 
Do tot have any example how you passed the payloads for the effects?! Eg using ZHA toolkit
Best Eric

**Comment by enyone** [comment]
@erkr I did use: Home Assistant -> Manage Zigbee device (ZHA) -> Clusters -> Commands -> 0x00FC that invoked this code statement https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6#file-lidltuyaeffects-py-L91 where there is this custom payload byte array https://gist.github.com/enyone/e1e0182f41ab9924d33fd24d77765aa6#file-lidltuyaeffects-py-L98

**Comment by erkr** [comment]
@enyone thanks, I will give that a try 😎
Ps: the payload encoding seems to be explained [here](https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3716#issuecomment-735004813)

**Comment by erkr** [comment]
@enyone 

I tried this, but it doesn't seem to do anything. 
I need to alter the `effect_id` slider first to enable the `issue zigbee command`, no errors but also no change. Not in effect mode, nor in static mode (adjust manual via the buttons on the box)

Any ideas?!

Best Eric

**Comment by Hedda** [comment]
> It'll also be hard to implement the effects properly and ZHA (HA Core) would probably also need to be modified for this.

For reference, there's a suggested draft implementation with prerequisite change to zha to allow stuff like this via quirks in the future:

* https://github.com/zigpy/zha/pull/762

Not quite a solution but that is at least a suggested prerequisite change to zha which could allow native control of effect commands via quirks in the future. But note that is just an example though and not actually an implementation, because as I understand that would then have to be done later as follow-up quirks to the [zha-device-handlers](https://github.com/zigpy/zha-device-handlers) repo after [zigpy/zha#762](https://github.com/zigpy/zha/pull/762) has been merged.

> _**As a test implementation, I've added support for the manufacturer-specific Hue lighting frame format in a fork of quirks. It look something like this (100% slop that manages to run)**_

FYI, also see related request/discussion for zha asking for ZHA to support "Hue native control" Dynamic Scenes Effects / Animations:

* https://github.com/zigpy/zha/issues/756

As well as equest/discussion for zha-device-handlers quirk handler for Philips Hue Dynamic Scenes Effects / Animations control:

* https://github.com/zigpy/zha-device-handlers/issues/2517

PS: I also think it is relevant to Open Home Foundation's roadmap discussion about shaping ZHA to have feature-parity with Z2M:

* https://github.com/OpenHomeFoundation/roadmap/issues/93

**Comment by ptc** [comment]
I got _TZE200_s8gkrkxk / TS0601 working with ZHA, including color, color temperature and all 16 effects.
Custom quirk and details: https://github.com/ptc/Melinera_HG06467

**Comment by Hedda** [comment]
> I got _TZE200_s8gkrkxk / TS0601 working with ZHA, including color, color temperature and all 16 effects.
> Custom quirk and details: https://github.com/ptc/Melinera_HG06467

Nice! Not sure they still sell this as a seasonal item at Lidl/Melinera but would be nice if you could try submit it as a pull request to this repo:

* https://github.com/zigpy/zha-device-handlers/pulls

See:

* https://github.blog/developer-skills/github/beginners-guide-to-github-creating-a-pull-request/

  * https://docs.github.com/en/pull-requests/how-tos/create-pull-requests/creating-a-pull-request

    * https://docs.github.com/en/pull-requests/reference/pull-requests

If you have a paid subscription to am AI/LLM then you could use it as a coding agent and have it guide you through the PR process:

* https://github.com/zigpy/zha-device-handlers/blob/dev/.github/copilot-instructions.md

Just make sure follow their newly published AI Policy:

* https://developers.home-assistant.io/blog/2026/07/20/ai-policy/

  * https://developers.home-assistant.io/docs/ai_policy/

The process is relatively straight-forward:

 - Fork the [zha-device-handler](https://github.com/zigpy/zha-device-handlers)s repo.
 - In your fork, create a new branch and write the code for your device to the correct file, (first adding the file if it does not exist).
 - Test to make sure it works.
 - From your branch with the changes in your fork, create a Pull Request against the [**dev**](https://github.com/zigpy/zha-device-handlers/tree/dev) branch of zha-device-handlers.

**Comment by erkr** [comment]
> I got _TZE200_s8gkrkxk / TS0601 working with ZHA, including color, color temperature and all 16 effects.
> Custom quirk and details: https://github.com/ptc/Melinera_HG06467

Thanks, it works great with my Christmas LED string 🎉
